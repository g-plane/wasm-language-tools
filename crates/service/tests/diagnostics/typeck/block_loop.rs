use super::*;
use insta::assert_json_snapshot;
use wat_service::LanguageService;

#[test]
fn folded() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func (param i32 f32) (result i32)
    (loop (result i32)
      (i32.add
        (local.get 0)
        (local.get 1)))))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn sequence() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func (param i32 f32) (result i32)
    loop (result i32)
      local.get 0
      local.get 1
      i32.add
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn results_folded() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func (result i32 i32)
    (loop (result i32 i32)
      (i32.const 0)))
  (func (result i32 i32)
    (loop (result i32 i32)
      (i32.const 0)
      (f32.const 0)))
  (func (result i32 i32)
    (loop (result i32 i32)
      (i32.const 0)
      (i32.const 0)
      (i32.const 0)))
  (func (result i32 i32)
    (loop (result i32 i32)
      (i32.const 0)
      (i32.const 0)
      (f32.const 0))))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn results_sequence() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func (result i32 i32)
    loop (result i32 i32)
      i32.const 0
    end)
  (func (result i32 i32)
    loop (result i32 i32)
      i32.const 0
      f32.const 0
    end)
  (func (result i32 i32)
    loop (result i32 i32)
      i32.const 0
      i32.const 0
      i32.const 0
    end)
  (func (result i32 i32)
    loop (result i32 i32)
      i32.const 0
      i32.const 0
      f32.const 0
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn results_correct() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func (result i32 i32)
    (loop (result i32 i32)
      unreachable))
  (func (result i32 i32)
    (loop (result i32 i32)
      (i32.const 0)
      (f32.const 0)
      (unreachable)))
  (func (result i32 i32)
    (loop (result i32 i32)
      (f32.const 0)
      (i32.const 0)
      (unreachable)))
  (func (result i32 i32)
    (loop (result i32 i32)
      (i32.const 0)
      (i32.const 0)))
  (func (result i32 i32)
    loop (result i32 i32)
      unreachable
    end)
  (func (result i32 i32)
    loop (result i32 i32)
      i32.const 0
      f32.const 0
      unreachable
    end)
  (func (result i32 i32)
    loop (result i32 i32)
      f32.const 0
      i32.const 0
      unreachable
    end)
  (func (result i32 i32)
    loop (result i32 i32)
      i32.const 0
      i32.const 0
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(response.items.is_empty());
}
