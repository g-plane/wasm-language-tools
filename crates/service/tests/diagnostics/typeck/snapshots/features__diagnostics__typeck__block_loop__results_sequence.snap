---
source: crates/service/tests/diagnostics/typeck/block_loop.rs
expression: response
---
{
  "kind": "full",
  "items": [
    {
      "range": {
        "start": {
          "line": 5,
          "character": 4
        },
        "end": {
          "line": 5,
          "character": 7
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [i32, i32], found [i32] at the end"
    },
    {
      "range": {
        "start": {
          "line": 10,
          "character": 4
        },
        "end": {
          "line": 10,
          "character": 7
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [i32, i32], found [i32, f32] at the end",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 9,
                "character": 6
              },
              "end": {
                "line": 9,
                "character": 17
              }
            }
          },
          "message": "expected type `i32`, found `f32`"
        }
      ]
    },
    {
      "range": {
        "start": {
          "line": 16,
          "character": 4
        },
        "end": {
          "line": 16,
          "character": 7
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [i32, i32], found [i32, i32, i32] at the end"
    },
    {
      "range": {
        "start": {
          "line": 22,
          "character": 4
        },
        "end": {
          "line": 22,
          "character": 7
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [i32, i32], found [i32, i32, f32] at the end",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 21,
                "character": 6
              },
              "end": {
                "line": 21,
                "character": 17
              }
            }
          },
          "message": "expected type `i32`, found `f32`"
        }
      ]
    }
  ]
}
